File: backend/commons/src/main/java/com/thinkus/commons/validation/CedulaECValidator.java
package com.thinkus.commons.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

/**
 * Validador de la anotación {@link CedulaEC}.
 *
 * <p>Implementa el algoritmo de Módulo 10 oficial del Registro Civil del Ecuador
 * para validar el dígito verificador de cédulas de personas naturales.
 *
 * <p>Algoritmo paso a paso:
 * <ol>
 *   <li>Verificar que el input sea exactamente 10 dígitos numéricos.</li>
 *   <li>Verificar que los dos primeros dígitos (provincia) estén entre 01 y 24.</li>
 *   <li>Verificar que el tercer dígito sea menor que 6 (persona natural).</li>
 *   <li>Multiplicar cada uno de los primeros 9 dígitos por los coeficientes
 *       [2, 1, 2, 1, 2, 1, 2, 1, 2].</li>
 *   <li>A cada producto, si es mayor que 9, restarle 9 (equivalente a sumar sus dígitos).</li>
 *   <li>Sumar todos los resultados.</li>
 *   <li>El dígito verificador esperado es: {@code (10 - (suma % 10)) % 10}.</li>
 *   <li>Validar que coincida con el décimo dígito de la cédula.</li>
 * </ol>
 *
 * <p>Esta clase es <b>thread-safe</b> e inmutable: no mantiene estado entre validaciones.
 *
 * <p>Devuelve {@code true} si el valor es {@code null} o vacío. La obligatoriedad
 * debe manejarse con {@code @NotBlank} si corresponde.
 */
public class CedulaECValidator implements ConstraintValidator<CedulaEC, String> {

    private static final int[] COEFICIENTES = {2, 1, 2, 1, 2, 1, 2, 1, 2};
    private static final int LONGITUD_CEDULA = 10;
    private static final int PROVINCIA_MIN = 1;
    private static final int PROVINCIA_MAX = 24;
    private static final int TERCER_DIGITO_MAX = 6;

    @Override
    public boolean isValid(String cedula, ConstraintValidatorContext context) {
        // null y vacío se consideran válidos: la obligatoriedad se valida con @NotBlank
        if (cedula == null || cedula.isBlank()) {
            return true;
        }

        String trimmed = cedula.trim();

        // 1. Longitud exacta de 10 caracteres
        if (trimmed.length() != LONGITUD_CEDULA) {
            return false;
        }

        // 2. Todos los caracteres deben ser dígitos
        for (int i = 0; i < LONGITUD_CEDULA; i++) {
            if (!Character.isDigit(trimmed.charAt(i))) {
                return false;
            }
        }

        // 3. Provincia entre 01 y 24
        int provincia = Integer.parseInt(trimmed.substring(0, 2));
        if (provincia < PROVINCIA_MIN || provincia > PROVINCIA_MAX) {
            return false;
        }

        // 4. Tercer dígito menor que 6 (cédula de persona natural)
        int tercerDigito = Character.getNumericValue(trimmed.charAt(2));
        if (tercerDigito >= TERCER_DIGITO_MAX) {
            return false;
        }

        // 5. Algoritmo Módulo 10 sobre los primeros 9 dígitos
        int suma = 0;
        for (int i = 0; i < COEFICIENTES.length; i++) {
            int digito = Character.getNumericValue(trimmed.charAt(i));
            int producto = digito * COEFICIENTES[i];
            suma += (producto > 9) ? producto - 9 : producto;
        }

        // 6. Cálculo del dígito verificador esperado
        int verificadorEsperado = (10 - (suma % 10)) % 10;
        int verificadorReal = Character.getNumericValue(trimmed.charAt(9));

        return verificadorEsperado == verificadorReal;
    }
}

File: frontend/src/app/shared/validators/cedula-ec.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * Validador reactivo de cedula ecuatoriana (algoritmo Modulo 10).
 *
 * Mismo algoritmo que aplica el backend en commons/CedulaECValidator.java.
 * Tener la validacion ESPEJO en frontend evita que el usuario llegue al
 * backend con una cedula obviamente mal formada (ahorra el round-trip).
 *
 * Reglas:
 *  1. Exactamente 10 digitos numericos.
 *  2. Provincia (primeros 2 digitos) entre 01 y 24.
 *  3. Tercer digito menor que 6 (cedulas de personas naturales).
 *  4. Coeficientes [2,1,2,1,2,1,2,1,2] sobre los primeros 9 digitos.
 *  5. Si el producto > 9, restarle 9.
 *  6. Sumar todos.
 *  7. Verificador esperado = (10 - suma % 10) % 10.
 *
 * Devuelve null si vacio (la obligatoriedad la maneja Validators.required).
 */
const COEFICIENTES = [2, 1, 2, 1, 2, 1, 2, 1, 2];

export function cedulaEcuatorianaValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const v = (control.value ?? '').toString().trim();

    if (!v) return null;

    if (!/^\d{10}$/.test(v)) {
      return { cedula: 'Debe contener exactamente 10 digitos' };
    }

    const provincia = parseInt(v.substring(0, 2), 10);
    if (provincia < 1 || provincia > 24) {
      return { cedula: 'Provincia invalida (debe estar entre 01 y 24)' };
    }

    if (parseInt(v[2], 10) >= 6) {
      return { cedula: 'Tercer digito invalido para persona natural' };
    }

    let suma = 0;
    for (let i = 0; i < COEFICIENTES.length; i++) {
      const producto = parseInt(v[i], 10) * COEFICIENTES[i];
      suma += producto > 9 ? producto - 9 : producto;
    }

    const verificador = (10 - (suma % 10)) % 10;
    if (verificador !== parseInt(v[9], 10)) {
      return { cedula: 'Digito verificador incorrecto' };
    }

    return null;
  };
}
